src/main/java/net/glowstone/block/blocktype/BlockRails.java

```diff
@@ -3,50 +3,443 @@
 import org.bukkit.block.BlockFace;
 import org.bukkit.inventory.ItemStack;
-import org.bukkit.material.MaterialData;
-import org.bukkit.material.Rails;
 import org.bukkit.util.Vector;
 
-import java.util.Arrays;
-import java.util.Collection;
+import java.util.*;
 
 public class BlockRails extends BlockNeedsAttached {
     @Override
     public Collection<ItemStack> getDrops(GlowBlock block, ItemStack tool) {
         return Arrays.asList(new ItemStack(block.getType()));
     }
 
-    private static BlockFace blockFaceFromYaw(float yaw) {
-        // nb: opposite from getOppositeBlockFace in BlockType
-        yaw = yaw % 360;
-        if (yaw < 0) {
-            yaw += 360.0;
+    @Override
+    public void placeBlock(GlowPlayer player, GlowBlockState state, BlockFace face, ItemStack holding, Vector clickedLoc) {
+        super.placeBlock(player, state, face, holding, clickedLoc);
+
+        GlowBlock block = state.getBlock();
+        RailDirection direction = getRailDirection(block);
+
+        state.setRawData((byte) direction.ordinal());
+        state.update(true, false);
+
+        updateNeighbors(block, direction);
+    }
+
+    private void updateNeighbors(GlowBlock block, RailDirection direction) {
+        direction.getNeighborRails(block).forEach(rail -> {
+            // check if the rail is stable
+            if (isUnstable(rail, block) && !isConnected(rail, block)) {
+                rail.setData((byte) getRailDirection(rail).ordinal());
+            }
+        });
+    }
+
+    private static boolean isConnected(GlowBlock rail, GlowBlock neighborRail) {
+        return RailDirection.getRailDirection(rail).getNeighborRails(rail).contains(neighborRail)
+                && RailDirection.getRailDirection(neighborRail).getNeighborRails(neighborRail).contains(rail);
+    }
+
+    private static boolean isUnstable(GlowBlock block, GlowBlock neighborRail) {
+        if (!isRailBlock(block)) {
+            return false;
+        }
+
+        Set<GlowBlock> neighborRails = RailDirection.getRailDirection(block).getNeighborRails(block);
+
+        // equals because blocks at the same location are not the same at the moment.
+        // TODO: !rail.equals(neighborRail) -> rail != neighborRail
+        return neighborRails.stream().filter(rail -> !rail.equals(neighborRail) && isConnected(block, rail)).count() != 2;
+    }
+
+    private static RailDirection getRailDirection(GlowBlock rail) {
+        // north - 0, south - 1, east - 2, west - 3, ascending_north - 4, ascending_south - 5, ascending_east - 6, ascending_west - 7
+        boolean[] unconnectedRails = new boolean[8];
+
+        GlowBlock north = rail.getRelative(BlockFace.NORTH);
+        GlowBlock south = rail.getRelative(BlockFace.SOUTH);
+        GlowBlock east = rail.getRelative(BlockFace.EAST);
+        GlowBlock west = rail.getRelative(BlockFace.WEST);
+
+        unconnectedRails[0] = isUnstable(north, rail) || isUnstable(north.getRelative(BlockFace.DOWN), rail);
```

Comment:
Since you are using magic values here (array index) for directions, perhaps you could name them?
e.g. `private static final int NORTH = 0;` and then use `unconnectedRails[NORTH] = isUnstable(north, rail) [...]`